### PR TITLE
feat: add proxy endpoint to update tenants

### LINKED ISSUE #47: Add proxy-endpoint to update tenants
Add a new `/api/tenants/update` to proxy to the `GET/POST/DELETE /api/update` endpoint of the `tenant service`.

### PR DESCRIPTION
**Pull request by xcoulon**
add proxy endpoints for the show/start/stop tenant updates
endpoints on the tenant service, and "audit log"
all requests along with their optional parameters (when
starting a new tenant update)

also, refactor "user search" test to use gock instead of go-vcr
(as the tenant update controller use as well)

also, add new event types in the DB

also, include the `authorization` header in the design
so it can be specified in the test requests

also, change default port for admin-console service
(when running locally) to `5345` to avoid colliding
with `cluster` service

Fixes #47

Signed-off-by: Xavier Coulon <xcoulon@redhat.com>

### CODE REVIEW

File: openshift/admin-console.app.yaml [hunk 1]
```diff
@@ -74,6 +74,11 @@ objects:
               configMapKeyRef:
                 name: admin-console
                 key: auth.url
+          - name: ADMIN_TENANT_URL
+            valueFrom:
+              configMapKeyRef:
+                name: admin-console
+                key: tenant.url
```
Comment (alexeykazakov) [inline openshift/admin-console.app.yaml]: Also add it to https://github.com/fabric8-services/admin-console/blob/master/openshift/admin-console.config.yaml

**Comment by alexeykazakov** [comment]
Don't forget to create config maps for tenant url in preview and prod.

**Comment by xcoulon** [comment]
thanks for the review @alexeykazakov ! I also created the config map update ticket yesterday (https://gitlab.cee.redhat.com/dtsd/housekeeping/issues/2516)
However, I realized this morning when waking up that I need to convert the request URI from `/api/tenant/updates` to `/api/update` when proxying the client request.

File: controller/tenants_updates_blackbox_test.go [hunk 2]
```diff
@@ -0,0 +1,341 @@
...
+	resource.Require(t, resource.Database)
+	config := configuration.New()
+	suite.Run(t, &TenantUpdatesControllerBlackboxTestSuite{DBTestSuite: testsuite.NewDBTestSuite(config)})
+}
+
+func (s *TenantUpdatesControllerBlackboxTestSuite) TestShowTenantUpdates() {
+	// given
+	config := testconfig.NewTenantUpdatesControllerConfigurationMock(s.T())
+	config.GetTenantServiceURLFunc = func() string {
+		return "https://test-tenant"
```
Comment (MatousJobanek) [inline controller/tenants_updates_blackbox_test.go]: the tenant host shouldn't contain `https://` scheme - it should be just `f8tenant`

File: controller/tenant_updates.go [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+		IdentityID:  identityID,
+		EventParams: auditlog.EventParams{},
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		return appl.AuditLogs().Create(ctx, &record)
+	})
+	if err != nil {
+		return app.JSONErrorResponse(ctx, err)
+	}
+	return httpsupport.RouteHTTP(ctx, c.config.GetTenantServiceURL())
```
Comment (MatousJobanek) [inline controller/tenant_updates.go]: let me clarify one thing - this endpoint is using `/tenant/updates` path and the `c.config.GetTenantServiceURL()` returns just `f8tenant` - shouldn't we specify also the target path in the tenant service here: `/api/update` ?

File: controller/tenants_updates_blackbox_test.go [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+	s.T().Run("ok", func(t *testing.T) {
+		// given
+		ctx, identity, err := testauth.EmbedUserTokenInContext(context.Background(), testauth.NewIdentity())
+		require.NoError(t, err)
+		tk := goajwt.ContextJWT(ctx)
+		require.NotNil(t, tk)
+		authzHeader := fmt.Sprintf("Bearer %s", tk.Raw)
+		gock.New("https://test-tenant").
+			Get("/api/tenant/updates").
```
Comment (MatousJobanek) [inline controller/tenants_updates_blackbox_test.go]: here, you should mock the endpoint in tenant service - `/api/update` not the endpoint in admin-proxy

File: controller/tenants_updates_blackbox_test.go [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+
+		t.Run("single clusters single env", func(t *testing.T) {
+			// given
+			ctx, identity, err := testauth.EmbedUserTokenInContext(context.Background(), testauth.NewIdentity())
+			require.NoError(t, err)
+			tk := goajwt.ContextJWT(ctx)
+			require.NotNil(t, tk)
+			authzHeader := fmt.Sprintf("Bearer %s", tk.Raw)
+			gock.New("https://test-tenant").
+				Post("/api/tenant/updates").
```
Comment (MatousJobanek) [inline controller/tenants_updates_blackbox_test.go]: you could also match the params (cluster url and env) using the `MatchParam` method

File: controller/tenant_updates.go (hunk 3 above)
Comment (MatousJobanek) [reply controller/tenant_updates.go]: didn't noticed that you have already realized it :-) https://github.com/fabric8-services/admin-console/pull/48#issuecomment-454300791

File: controller/tenants_updates_blackbox_test.go (hunk 2 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: ok, let me change that

File: controller/tenant_updates.go (hunk 3 above)
Comment (xcoulon) [reply controller/tenant_updates.go]: yes, that's what I need to change as well (see https://github.com/fabric8-services/admin-console/pull/48#issuecomment-454300791)

File: controller/tenants_updates_blackbox_test.go (hunk 4 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: yes, that's also what I need to change as a consequence of https://github.com/fabric8-services/admin-console/pull/48#issuecomment-454300791

File: controller/tenants_updates_blackbox_test.go (hunk 5 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: ok, good point, let me do that

File: design/tenant_updates.go [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+var _ = a.Resource("tenant_updates", func() {
+	a.BasePath("/tenant/updates")
+
+	a.Action("show", func() {
+		a.Security("jwt")
+		a.Routing(
+			a.GET(""),
+		)
+		a.Headers(func() {
+			a.Header("Authorization", d.String, "the authorization header")
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply design/tenant_updates.go]: because I need to be able to specify the `Authorization` header in the test, and with this `Header` func, the generated method to test the endpoint will provide me with an arg to pass the value. See https://github.com/fabric8-services/admin-console/pull/48/files#diff-a2bd8cd90ce50fc5463d8b3539620c92R83

File: controller/tenant_updates.go [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+}
+
+// Start starts a tenant update
+func (c *TenantUpdatesController) Start(ctx *app.StartTenantUpdatesContext) error {
+	tokenManager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
+	}
+	identityID, err := tokenManager.Locate(ctx)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 8]
```diff
@@ -0,0 +1,111 @@
...
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid authorization token (invalid 'sub' claim)"))
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 9]
```diff
@@ -0,0 +1,111 @@
...
+	record := auditlog.AuditLog{
+		EventTypeID: auditlog.ShowTenantUpdate,
+		IdentityID:  identityID,
+		EventParams: auditlog.EventParams{},
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		return appl.AuditLogs().Create(ctx, &record)
+	})
+	if err != nil {
+		return app.JSONErrorResponse(ctx, err)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 10]
```diff
@@ -0,0 +1,111 @@
...
+
+// Show returns information about the ongoing tenant update
+func (c *TenantUpdatesController) Show(ctx *app.ShowTenantUpdatesContext) error {
+	tokenManager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
+	}
+	identityID, err := tokenManager.Locate(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid authorization token (invalid 'sub' claim)"))
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 11]
```diff
@@ -0,0 +1,111 @@
...
+	}
+}
+
+// Show returns information about the ongoing tenant update
+func (c *TenantUpdatesController) Show(ctx *app.ShowTenantUpdatesContext) error {
+	tokenManager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
+	}
+	identityID, err := tokenManager.Locate(ctx)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+	return httpsupport.RouteHTTP(ctx, c.config.GetTenantServiceURL())
+}
+
+// Stop stops the ongoing tenant update
+func (c *TenantUpdatesController) Stop(ctx *app.StopTenantUpdatesContext) error {
+	tokenManager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
+	}
+	identityID, err := tokenManager.Locate(ctx)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+	record := auditlog.AuditLog{
+		EventTypeID: auditlog.StartTenantUpdate,
+		IdentityID:  identityID,
+		EventParams: eventParams,
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		return appl.AuditLogs().Create(ctx, &record)
+	})
+	if err != nil {
+		return app.JSONErrorResponse(ctx, err)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+
+// Stop stops the ongoing tenant update
+func (c *TenantUpdatesController) Stop(ctx *app.StopTenantUpdatesContext) error {
+	tokenManager, err := token.ReadManagerFromContext(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
+	}
+	identityID, err := tokenManager.Locate(ctx)
+	if err != nil {
+		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("invalid authorization token (invalid 'sub' claim)"))
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenant_updates.go [hunk 15]
```diff
@@ -0,0 +1,111 @@
...
+	record := auditlog.AuditLog{
+		EventTypeID: auditlog.StopTenantUpdate,
+		IdentityID:  identityID,
+		EventParams: auditlog.EventParams{},
+	}
+	err = application.Transactional(c.db, func(appl application.Application) error {
+		return appl.AuditLogs().Create(ctx, &record)
+	})
+	if err != nil {
+		return app.JSONErrorResponse(ctx, err)
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 18e5a49

File: controller/tenants_updates_blackbox_test.go (hunk 5 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: done in 02cc202

File: controller/tenants_updates_blackbox_test.go (hunk 4 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: done in 02cc202

File: controller/tenant_updates.go (hunk 3 above)
Comment (xcoulon) [reply controller/tenant_updates.go]: done in 02cc202

File: controller/tenants_updates_blackbox_test.go (hunk 2 above)
Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: done in 02cc202 (using `http://`, but that should work as well, I believe)

File: openshift/admin-console.app.yaml (hunk 1 above)
Comment (xcoulon) [reply openshift/admin-console.app.yaml]: done in d277562

File: design/tenant_updates.go [hunk 16]
```diff
@@ -0,0 +1,61 @@
...
+		})
+		a.Description("Get information about last/ongoing update.")
+		a.Response(d.OK) // here we don't specify a media type, because we're just proxying to `auth`
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply design/tenant_updates.go]: done in 18e5a49

File: design/tenant_updates.go [hunk 17]
```diff
@@ -0,0 +1,61 @@
...
+		a.Params(func() {
+			a.Param("cluster_url", d.String, "the URL of the OSO cluster the update should be limited to")
+			a.Param("env_type", d.String, "environment type the update should be executed for", func() {
+				a.Enum("user", "che", "jenkins", "stage", "run")
+			})
+		})
+
+		a.Description("Start new cluster-wide update.")
+		a.Response(d.Accepted)
+		a.Response(d.Conflict) // here we don't specify a media type, because we're just proxying to `auth`
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply design/tenant_updates.go]: done in 18e5a49

File: controller/tenants_updates_blackbox_test.go (hunk 2 above)
Comment (MatousJobanek) [reply controller/tenants_updates_blackbox_test.go]: my point was that the host doesn't contain any scheme at all - it is  just `f8tenant`, so the whole URL would be: `f8tenant/api/update`

Comment (xcoulon) [reply controller/tenants_updates_blackbox_test.go]: the config map has been updated to include the `http://` scheme, so we should be good with that now

File: controller/search.go [hunk 18]
```diff
@@ -36,12 +37,11 @@ func NewSearchController(service *goa.Service, config SearchControllerConfigurat
...
-	tokenManager, err := token.ReadManagerFromContext(ctx)
-	if err != nil {
-		return app.JSONErrorResponse(ctx, errors.NewUnauthorizedError("missing token manager in the request context"))
-	}
-	identityID, err := tokenManager.Locate(ctx)
+	identityID, err := authsupport.LocateIdentity(ctx)
 	if err != nil {
+		log.Error(ctx, map[string]interface{}{
+			"err": err,
+		}, "unable to proxy to auth service")
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/search.go]: ok, let me rephrase those error messages

File: configuration/configuration.go [hunk 19]
```diff
@@ -307,6 +309,11 @@ func (c *Configuration) IsCleanTestDataEnabled() bool {
 	return c.v.GetBool(varCleanTestDataEnabled)
 }
 
+// IsCleanTestDataErrorReportingRequired returns `true` if the test data should be cleaned after each test. (default: true)
+func (c *Configuration) IsCleanTestDataErrorReportingRequired() bool {
```
[the remark this answers is not in the document]
Comment (xcoulon) [reply configuration/configuration.go]: that's required by the latest `fabric8-common`: https://github.com/fabric8-services/fabric8-common/blob/f06e14243602e383835a8f8cbf3675e24a738322/test/suite/db_clean.go#L88

File: controller/search.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (xcoulon) [reply controller/search.go]: done in 18d79ae

File: configuration/configuration.go (hunk 19 above)
[the remark this answers is not in the document]
Comment (MatousJobanek) [reply configuration/configuration.go]: ok, but is it good that we have a method/config var that is used only for the test purposes in the normal configuration file? shouldn't that be separated? I'm just asking ;-)

[the remark this answers is not in the document]
Comment (xcoulon) [reply configuration/configuration.go]: it's already the case for a few other flags that are only used for testing purpose, but I'm fine witj revisiting this in a separate PR

File: design/tenant_updates.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (alexeykazakov) [reply design/tenant_updates.go]: You don't really have to specify the Authorization header explicitly. You can just embed the token into the context as we do everywhere else. Am I missing something?

[the remark this answers is not in the document]
Comment (xcoulon) [reply design/tenant_updates.go]: Yes, I know this is what we usually do with our endpoint tests, but in this particular case it did not work: when testing the proxying with gork, there was no `Authorization` header so Gork would not match the HTTP request sent to the remote cluster service.
